### PR TITLE
Draw Trajectories onto X-Rays and ProbabilityGrids.

### LINKED ISSUE #174: Optionally draw trajectories on X-Rays

### PR DESCRIPTION
**Pull request by SirVer**
This behavior can be turned off with the 'draw_trajectories' setting in Lua.

Fixes #174.

### CODE REVIEW

**Comment by SirVer** [comment]
@damienrg We talked about this in a PR or bug recently, I went ahead and implemented this since I was dearly missing that feature.

**Comment by SirVer** [comment]
Sample output: 3D has 3 trajectories, 2D has 2 trajectories.

xray_xy_all
probability_grid_no_free

**Comment by damienrg** [comment]
> @damienrg We talked about this in a PR or bug recently, I went ahead and implemented this since I was dearly missing that feature.
No problem. Do you still want images in SVG or TIF?

File: cartographer/io/draw_trajectories.cc [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        pose_to_pixel(transform::ToRigid3(node.pose()));
+    cairo_line_to(cr.get(), pixel.x(), pixel.y());
+  }
+  cairo_stroke(cr.get());
+
+  // Draw beginning and end markers.
+  {
+    const Eigen::Array2i pixel =
+        pose_to_pixel(transform::ToRigid3(trajectory.node(0).pose()));
+    cairo_set_source_rgba(cr.get(), 0., 255., 0., kAlpha);
```
Comment (damienrg) [inline cartographer/io/draw_trajectories.cc]: replace 255. by 1.0

File: cartographer/io/draw_trajectories.cc [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    cairo_arc(cr.get(), pixel.x(), pixel.y(), kTrajectoryEndMarkers, 0,
+              2 * M_PI);
+    cairo_fill(cr.get());
+  }
+  {
+    const Eigen::Array2i pixel = pose_to_pixel(transform::ToRigid3(
+        trajectory.node(trajectory.node_size() - 1).pose()));
+    cairo_set_source_rgba(cr.get(), 255., 0., 0., kAlpha);
```
Comment (damienrg) [inline cartographer/io/draw_trajectories.cc]: replace:
cairo_set_source_rgba(cr.get(), 0., 255., 0., kAlpha);
by:
cairo_set_source_rgba(cr.get(), 0., 1., 0., kAlpha);

File: cartographer/io/draw_trajectories.cc [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  if (trajectory.node_size() == 0) {
+    return;
+  }
+  constexpr double kTrajectoryWidth = 4.;
+  constexpr double kTrajectoryEndMarkers = 6.;
+  constexpr double kAlpha = 0.7;
+
+  auto cr = ::cartographer::io::MakeUniqueCairoPtr(cairo_create(surface));
+
+  cairo_set_source_rgba(cr.get(), color[0] / 255., color[1] / 255.,
```
Comment (damienrg) [inline cartographer/io/draw_trajectories.cc]: If you apply #423 before remove `/ 255`

File: cartographer/io/draw_trajectories.cc (hunk 1 above)
Comment (damienrg) [inline cartographer/io/draw_trajectories.cc]: `cairo_set_source_rgba(cr.get(), 0., 1., 0., kAlpha);`

**Comment by SirVer** [comment]
@damienrg, Thanks for looking through the code!

> No problem. You implemented it by writing PNG, but do you still want images in SVG or TIF in the future?

Unfortunately I did not find a way of writing a layered SVG or TIF from Cairo. Without layers I did not see the benefit of another format, so I kept PNG for now. Having a layered image format would still be nice, but it would probably mean transitioning away from Cairo which is a bigger investment.

**Pull request merged**